File: Python/0133-clone-graph.py
# Time:  O(n)
# Space: O(n)
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None

        cloned_node = Node(node.val)

        cloned, stack = {node:cloned_node}, [node]

        while stack:
            cur = stack.pop()
            for neighbor in cur.neighbors:
                if neighbor not in cloned:
                    stack.append(neighbor)
                    cloned_neighbor = Node(neighbor.val)
                    cloned[neighbor] = cloned_neighbor
                cloned[cur].neighbors.append(cloned[neighbor])

        return cloned[node]







# Notes: Use dictionary cloned to search in constant time
# DFS runtime V+E , E could be O(V^2) in case of dense graph
# cloned dict contains visited and to be visited ie stack nodes

File: C++/0133-clone-graph.cpp
// Tags: DFS BFS Graph Amazon
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
// Time: O(V + E), E could be O(V^2) in case of dense graph
// Space: O(V + E)
class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (!node)
            return nullptr;
        unordered_map<Node*, Node*> cloned;
        Node* cnode = new Node(node->val);
        cloned[node] = cnode;
        stack<Node*> st;
        st.push(node);
        while (!st.empty()) {
            Node *cur = st.top(); st.pop();
            for (Node* neighbor : cur->neighbors) {
                if (!cloned.count(neighbor)) {
                    // ensure each node is visited only once
                    // (mark it visited)
                    cloned[neighbor] = new Node(neighbor->val);
                    st.push(neighbor);
                }
                cloned[cur]->neighbors.push_back(cloned[neighbor]);
            }
        }
        return cnode;
    }
};
/*
cloned Map contains visited and to be visited ie stack nodes

In dfs, we don't want to push the nodes which are either already
visited or in stack currently.
*/
